moto/sns/responses.py

```diff
@@ -78,6 +78,7 @@ def _transform_message_attributes(self, message_attributes):
                                 "when calling the Publish operation: "
                                 f"Could not cast message attribute '{name}' value to number."
                             )
+                    transform_value = str(transform_value)
```

Comment:
If we don't re-assign `transform_value` in the first place, I'm assuming this isn't necessary.

I.e.: instead of doing `transform_value = int(value["StringValue"])`, we can just try to cast (`int(value["StringValue"])`) without assigning it to `transform_value`. That way, we'll keep the original value as a string.